### ISSUE TITLE
emacsclient does not work in Nix shells due to changed TMPDIR

### ISSUE THREAD

**Issue by ejpcmac** [NixOS/nixpkgs#55201]
## Issue description

On NixOS, `TMPDIR` is not set by default, but it is set to `/run/user/$UID` in Nix shells.

If I set `services.emacs.enable = true`, an Emacs daemon is started automatically by `systemd`, using the system environment. This, as `TMPDIR` is not defined, the socket is created under `/tmp`.

If I use `emacsclient` outside of a Nix shell, all works properly. Now, if I use `emacsclient` *inside* a Nix shell, it looks for the socket under `/run/user/$UID` as `TMPDIR` is set: it does not work.

### Steps to reproduce

1. Start an Emacs daemon outside of a Nix shell
2. Start a Nix shell
3. Run `emacsclient -c`: it does not work

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.19.16, NixOS, 19.03pre166987.bc41317e243 (Koi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.2`
 - channels(root): `"home-manager, jpc-nix, nixos-19.03pre166987.bc41317e243"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

## Possible workarounds

1. Setting an `emacsclient` alias prefixed by `TPMDIR=/tmp`. This would work as expected, but it would be platform-specific: on macOS for instance, the `TMPDIR` is correctly set both inside and outside Nix shells.

2. Setting `environment.variables.TPMDIR = "/run/user/$UID";`. I went for this one, but I a now facing another issue: when running `nixos-rebuild switch` after a reboot, I get a `/run/user/0` not found error, and I need to manually create this directory.

Is there any reason `TMPDIR` is not set outside Nix shells?

**Comment by ejpcmac** [comment]
I still use the workaround from my first post, and it still feels hacky to define it this way.

**Comment by aij** [comment]
This annoyed me too when I first ran into it, though now I'm actually taking advantage of it to run separate emacs inside and outside of an FHS chroot via nix-shell. I could of course explicitly set environment variables for that use case. Making them consistent by default would certainly be less confusing.

**Comment by ejpcmac** [comment]
@aij Yes, I think the current behaviour is confusing. Nonetheless, I’m now using a shortcut in ranger to open Emacs in the current directory, and not `emacsclient` that often, so maybe I wouldn’t be hit by this issue so much in fact. Let’s try to unset my `TMPDIR` and see if I still have issues with my current workflow.

**Comment by ejpcmac** [comment]
Effectively, after removing my custom `TMPDIR` I’m not hit by this issue anymore, since my use case has evolved. Let’s say it’s a very specific issue that don’t affect many people, so we can close it and focus on more important things :)

**Comment by tobiasBora** [comment]
I actually had this exact issue, and I'm quite happy with my current workflow... Is there any reason why `nix-shell` is creating a custom `/tmp` directory? (maybe related with https://github.com/NixOS/nix/issues/395) If not what is the recommended fix? I guess that emacs is quite popular, so I'm not sure it's that specific...

**Comment by ejpcmac** [comment]
In fact I’m still hitting this issue myself when running `git flow` commands in a Nix shell: when it tries to open my `$EDITOR`, which is some `emacsclient` command, it does not find the server. As I’m not alone to encounter this issue, let’s reopen it!

I’ll take a look on the `programs.direnv.enableNixDirenvIntegration` option to see what’s going on. It does not seem to be available in the `release-19.09` branch for me to try it though—I cannot upgrade to NixOS 20.03 until September because I’m on a slow VSAT link at the moment and need to wait for the upgrades to come by the next postal mail… :smile:

**Comment by Blaisorblade** [comment]
> 1. Setting an `emacsclient` alias prefixed by `TPMDIR=/tmp`. This would work as expected, but it would be platform-specific: on macOS for instance, the `TMPDIR` is correctly set both inside and outside Nix shells.

I'm affected by the issue _on Mac_ using direnv — Macs default to a per-user TMPDIR in `/var/folders/...`, but Nix resets TMPDIR to `/tmp`. I see this both using `nix-shell` and when just using `direnv` to load the same configuration.

**Comment by pcasaretto** [comment]
I see the issue is solved for nix-direnv, but did you ever find a workaround for nix-shell (and by extension `nix develop`) @Blaisorblade ?

**Comment by Blaisorblade** [comment]
@pcasaretto Nope, and I stopped using Nix so I'm afraid I can't help :-(

**Comment by eqyiel** [comment]
> I'm affected by the issue _on Mac_ using direnv — Macs default to a per-user TMPDIR in `/var/folders/...`, but Nix resets TMPDIR to `/tmp`. I see this both using `nix-shell` and when just using `direnv` to load the same configuration.

Yeah, on macOS, setting `TMPDIR` to static `/tmp` doesn't work because the location of default `TMPDIR` is not static. I'm using the following workaround in a wrapper script to allow e.g. `emacsclient` to find the server file.

```console
TMPDIR="$(getconf DARWIN_USER_TEMP_DIR)"
```

## Example of the issue

I'm not sure how much of this is nix shell and how much is direnv, but this is what `TMPDIR` vs `getconf DARWIN_USER_TEMP_DIR` looks like in a nix shell env created by direnv's `use flake` on macOS:

```console
$ echo $IN_NIX_SHELL
impure

$ echo $TMPDIR
/var/folders/m4/gwvz5s8j6mg02fgcm7zr28100000gn/T/nix-shell.A1dFsu

$ getconf DARWIN_USER_TEMP_DIR
/var/folders/m4/gwvz5s8j6mg02fgcm7zr28100000gn/T/
```

Outside:

```
$ echo $TMPDIR
/var/folders/m4/gwvz5s8j6mg02fgcm7zr28100000gn/T/

$ getconf DARWIN_USER_TEMP_DIR
/var/folders/m4/gwvz5s8j6mg02fgcm7zr28100000gn/T/
```

https://github.com/direnv/direnv/issues/1057
